src/world.mojo

```diff
@@ -0,0 +1,1705 @@
+
+        Args:
+            components: The components to add to the entity.
+        """
+        self._check_locked()
+
+        alias size = components.__len__()
+
+        component_info = self._component_manager.get_info_arr[*Ts]()
+        archetype_index = self._get_archetype_index(component_info)
+        entity = self._create_entity(archetype_index)
+        index_in_archetype = self._entities[int(entity.id)].index
+
+        archetype = self._archetypes.get_ptr(archetype_index)
+
+        @parameter
+        for i in range(components.__len__()):
+            archetype[].unsafe_set(
+                int(index_in_archetype),
+                component_info[i].id,
+                UnsafePointer.address_of(components[i]).bitcast[UInt8](),
+            )
+
+        # TODO
+        # if self._listener != nil:
+        #     var newRel *Id
+        #     if arch.HasRelationComponent:
+        #         newRel = &arch.RelationComponent
+
+        #     var bits = subscription(true, false, len(comps) > 0, false, newRel != nil, newRel != nil)
+        #     var trigger = self._listener.Subscriptions() & bits
+        #     if trigger != 0 && subscribes(trigger, &arch.Mask, nil, self._listener.Components(), nil, newRel):
+        #         self._listener.Notify(self, EntityEventEntity: entity, Added: arch.Mask, AddedIDs: comps, NewRelation: newRel, EventTypes: bits)
+
+        return
+
+    fn set[T: ComponentType](inout self, entity: Entity, component: T) raises:
+        """
+        Overwrites a component for an [Entity], using the given pointer for the content.
+        """
+        entity_index = self._entities[int(entity.id)]
+        self._archetypes[int(entity_index.archetype_index)].set(
+            int(entity_index.index),
+            self._component_manager.get_ref(component),
+        )
+
+    fn set[
+        *Ts: ComponentType
+    ](inout self, entity: Entity, *components: *Ts) raises:
+        """
+        Overwrites a component for an [Entity], using the given pointer for the content.
+        """
+        entity_index = self._entities[int(entity.id)]
+        archetype = self._archetypes.get_ptr(int(entity_index.archetype_index))
+
+        @parameter
+        for i in range(components.__len__()):
+            archetype[].set(
+                int(entity_index.index),
+                self._component_manager.get_ref(components[i]),
+            )
+
+    fn get[
+        T: ComponentType
+    ](inout self, entity: Entity) raises -> ref [self._archetypes[0]] T:
+        """Returns a reference to the given component of an [Entity].
+
+        Raises:
+            Error: If the entity does not have the component.
+        """
+        entity_index = self._entities[int(entity.id)]
+        return (
+            self._archetypes[int(entity_index.archetype_index)]
+            .get_component_ptr(
+                int(entity_index.index),
+                self._component_manager.get_id[T](),
```

Comment:
I am not sure I understand this fully but would like to.
I think this could be solved without dict lookups. I will implement a constant-time fast lookup table for components with small IDs soon. Later, this could be solved with class variables of the component types (which then would need to inherit from some base-component type). Neither inheritence nor class variables exist at the moment, though.

Reply:
Not sure how you want to implement a mapping from a type to its ID faster than a Dict. The other way round, where the ID is the key, sure... but this way?

The class variable could be a solution. But as it is now, it is certainly waaay slower than it could be.